src/Common/src/TypeSystem/Common/TypeSystemContext.cs

```diff
@@ -736,5 +736,42 @@ protected internal virtual bool ComputeHasGCStaticBase(FieldDesc field)
             // Type system contexts that support this need to override this.
             throw new NotSupportedException();
         }
+
+        /// <summary>
+        /// Extension point for computation of type flags. This extension point may be overriden to customize
+        /// type flags that are not computed by the ComputeTypeFlags api on TypeDesc.
+        /// At this time, the flag so computed are:
+        /// HasStaticConstructor
+        /// 
+        /// When overriding, return the value of calling this base implementation passing the flags value computed
+        /// in the override.
+        /// </summary>
+        protected internal virtual TypeFlags ComputeTypeFlags(TypeDesc type, TypeFlags flags, TypeFlags mask)
+        {
+            // If we are looking to compute HasStaticConstructor, and we haven't yet assigned a value
+            if (((mask & TypeFlags.HasStaticConstructorComputed) == TypeFlags.HasStaticConstructorComputed) &&
+                ((flags & TypeFlags.HasStaticConstructorComputed) == 0))
+            {
+                MetadataType metadataType = type.GetTypeDefinition() as MetadataType;
```

Comment:
Nit: you could avoid the cast to `MetadataType` for the generic instantiation case and it might improve readability too.